src/deserialize/parse.rs

```diff
@@ -0,0 +1,361 @@
+        .collect::<String>()
+        .len();
+    let mut number = String::new();
+    let string = chars.take(c_len).collect::<String>();
+    number.push(n);
+    number.push_str(&string);
+
+    match number {
+        n if n.parse::<usize>().is_ok() => Ok(Edn::UInt(n.parse::<usize>()?)),
+        n if n.parse::<isize>().is_ok() => Ok(Edn::Int(n.parse::<isize>()?)),
+        n if n.parse::<f64>().is_ok() => Ok(Edn::Double(n.parse::<f64>()?.into())),
+        n if n.contains("/") && n.split("/").all(|d| d.parse::<f64>().is_ok()) => {
+            Ok(Edn::Rational(n))
+        }
+        _ => Err(Error::ParseEdn(format!("{} could not be parsed", number))),
+    }
+}
+
+fn read_char(chars: &mut std::str::Chars) -> Result<Edn, Error> {
+    let c = chars.next();
+    c.ok_or(format!("{:?} could not be parsed", c))
+        .map(|c| Edn::Char(c))
+        .map_err(|e| Error::ParseEdn(e))
+}
+
+fn read_bool_or_nil(c: char, chars: &mut std::str::Chars) -> Result<Edn, Error> {
+    match c {
+        't' => {
+            let c_len = chars
+                .clone()
+                .take_while(|e| e == &'r' || e == &'u' || e == &'e')
+                .collect::<String>()
+                .len();
+            let mut string = String::new();
+            let t = chars.take(c_len).collect::<String>();
+            string.push(c);
+            string.push_str(&t);
+            Ok(Edn::Bool(string.parse::<bool>()?))
+        }
+        'f' => {
+            let c_len = chars
+                .clone()
+                .take_while(|e| e == &'a' || e == &'l' || e == &'s' || e == &'e')
+                .collect::<String>()
+                .len();
+            let mut string = String::new();
+            let f = chars.take(c_len).collect::<String>();
+            string.push(c);
+            string.push_str(&f);
+            Ok(Edn::Bool(string.parse::<bool>()?))
+        }
+        'n' => {
+            let c_len = chars
+                .clone()
+                .take_while(|e| e == &'i' || e == &'l')
+                .collect::<String>()
+                .len();
+            let mut string = String::new();
+            let n = chars.take(c_len).collect::<String>();
+            string.push(c);
+            string.push_str(&n);
+            match &string[..] {
+                "nil" => Ok(Edn::Nil),
+                _ => Err(Error::ParseEdn(format!("{} cound not be parsed", string))),
+            }
+        }
+        _ => Err(Error::ParseEdn(
+            "Nullable boolean cound not be parsed".to_string(),
+        )),
+    }
+}
+
+fn read_vec(chars: &mut std::str::Chars) -> Result<Edn, Error> {
+    let mut res: Vec<Edn> = vec![];
+    loop {
+        match chars.next() {
+            Some(']') => return Ok(Edn::Vector(Vector::new(res))),
+            Some(c) if !c.is_whitespace() && c != ',' => {
+                res.push(parse(Some(c), chars)?);
+            }
+            Some(c) if c.is_whitespace() || c == ',' => (),
```

Comment:
I think we can make those two `Some(c)` arms faster by calling `is_whitespace` and checking if the character is a `,` only once.

Maybe something like this:
```rust
Some(c) => {
    let is_comma = c == ',';
    if !c.is_whitespace() && !is_comma {
        res.push(parse(Some(c), chars)?);
    }
}
```

Another option is to:
```rust
Some(c) if !c.is_whitespace() && c != ',' => {
    res.push(parse(Some(c), chars)?);
}
```
And to just remove the second `Some(c)` since it does nothing :wink:

Reply:
You cant remove the second one because he is responsible for skipping values.

Reply:
they both should exist because it seems important to have an `err` case. If we get something unknown we should return an error and not just skip it.

Reply:
Tried replacing for
```
const fn should_skip_char(c: &char) -> bool {
    *c == ' ' || *c == ',' || *c == '\n' || *c == '\r' || *c == '\t'
}
```

and this:

```
const fn should_skip_char(c: &char) -> bool {
     match *c {
         ' ' | ',' | '\x09'..='\x0d' => true,
         _ => false
     }
}
```

And it only worsen performance

Reply:
This also worsened performance:
```
let ch = chars.next();
        let should_skip = ch.unwrap_or(' ').is_whitespace() || ch.unwrap_or(' ') == ',';
        match ch {
            Some('}') => return Ok(Edn::Map(Map::new(res))),
            Some(c) if !should_skip => {
                if key.is_some() {
                    val = Some(parse(Some(c), chars)?);
                } else {
                    key = Some(parse(Some(c), chars)?);
                }
            }
            Some(_) if should_skip => (),
            err => return Err(Error::ParseEdn(format!("{:?} could not be parsed", err))),
        }
```

Reply:
This actually worsen performance by 2us:
```
match c.next() {
            Some(']') => return Ok(Edn::Vector(Vector::new(res))),
            Some(c) => {
                let should_skip = c.is_whitespace() || c == ',';
                if !should_skip {
                    res.push(parse(Some(c), chars)?);
                }
            }
}```

Reply:
How about using this if then on the `Err` value?

Reply:
it doesn't improve performance as the other cases show